src/code/z_horse.c

```diff
@@ -1,21 +1,235 @@
+// unused
+s32 Horse_CopyPointFromPathList(PlayState* play, s32 pathIndex, s32 pointIndex, Vec3s* dst, s16* arg4) {
+    Path* path = &play->setupPathList[pathIndex];
+    Vec3s* points;
+    s32 count = path->count;
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_horse/func_800F3ED4.s")
+    dst->x = 0;
+    dst->y = 0;
+    dst->z = 0;
+    *arg4 = 0;
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_horse/func_800F40A0.s")
+    if (count == 0) {
+        return false;
+    }
+    if (pointIndex >= count) {
+        return false;
+    }
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_horse/func_800F415C.s")
+    points = Lib_SegmentedToVirtual(path->points);
+    points += pointIndex;
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_horse/func_800F41E4.s")
+    dst->x = points->x;
+    dst->y = points->y;
+    dst->z = points->z;
+    *arg4 = 0;
+
+    return true;
+}
+
+typedef struct HorseValidScene {
+    /* 0x0 */ s16 sceneId;
+    /* 0x2 */ s16 sceneLayerMinusOne;
+} HorseValidScene; // size = 0x4
+
+HorseValidScene sHorseValidScenes[] = {
+    { SCENE_00KEIKOKU, 0 },      // Termina Field
+    { SCENE_24KEMONOMITI, 0 },   // Road to southern swap
+    { SCENE_F01, 0 },            // Romani ranch
+    { SCENE_KOEPONARACE, 0 },    // Gorman track
+    { SCENE_20SICHITAI, 0 },     // Southern swamp poisoned
+    { SCENE_20SICHITAI2, 0 },    // Souther swamp clear
+    { SCENE_30GYOSON, 0 },       // Great bay coast
+    { SCENE_31MISAKI, 0 },       // Zora cape
+    { SCENE_ROMANYMAE, 0 },      // Milk road
+    { SCENE_IKANAMAE, 0 },       // Road to Ikana
+    { SCENE_13HUBUKINOMITI, 0 }, // Path to Mountain Village
+};
+
+s32 Horse_IsValidSpawn(s16 sceneId) {
+    s32 i;
+
+    for (i = 0; i < ARRAY_COUNT(sHorseValidScenes); i++) {
+        if (sceneId == sHorseValidScenes[i].sceneId) {
+            return true;
+        }
+    }
+
+    return false;
+}
+
+void Horse_ResetHorseData(PlayState* play) {
+    gSaveContext.save.saveInfo.horseData.sceneId = SCENE_F01;
+    gSaveContext.save.saveInfo.horseData.pos.x = -1420;
+    gSaveContext.save.saveInfo.horseData.pos.y = 257;
+    gSaveContext.save.saveInfo.horseData.pos.z = -1285;
+    gSaveContext.save.saveInfo.horseData.yaw = 0x2AAA;
+}
+
+s32 gHorseIsMounted = false;
+s32 D_801BDAA0 = false;
+s32 gHorsePlayedEponasSong = false;
+
+HorseValidScene sHorseValidSceneLayers[] = {
+    { SCENE_00KEIKOKU, 5 - 1 },      // Termina Field - First Cycle
+    { SCENE_30GYOSON, 1 - 1 },       // Great Bay Coast - Post-Gyorg
+    { SCENE_31MISAKI, 1 - 1 },       // Zora Cape - Post-Gyorg
+    { SCENE_13HUBUKINOMITI, 1 - 1 }, // Path to Mountain Village - Post-Goht
+};
+
+s32 Horse_IsValidSceneLayer(PlayState* play, Player* player) {
+    s32 i;
+
+    // Layer 0 is normal gameplay
+    if (gSaveContext.sceneLayer == 0) {
+        return true;
+    }
+
+    // Layers above 0 are usually cutscenes, but some of them are actually natural playable layers
```

Comment:
sceneLayers aren't really tied to cutscenes anymore, unlike OoT, see here:
https://discord.com/channels/688807550715560050/688851976746041391/1058602699865731102

So this doesn't feel quite right. Layer 0 can be a cutscene with the right entrance, and higher layers can be normal gameplay. So I'm not really sure how to describe why `sceneLayer 0` gets an exception, maybe this was written before cutscenes were reworked? Idk

Reply:
This probably is a leftover from OoT and everybody just forgot to update it when they rewrote the cutscenes system.
We could speculate about this, but idk if we want a comment documentating just speculation about this

Reply:
I would agree, I think there is some correlation to cutscenes but 🤷. Perhaps just remove the existing comments since we are not sure?

Reply:
Yeah, I'd rather wait until we dig a little more into scene layers, so I'd also prefer to remove the comments for now?